### PR TITLE
#11 create class for parsing command

### PR DESCRIPTION
**Pull request by ksergey12**
connect to #11

### CODE REVIEW

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 1]
```diff
@@ -1,19 +1,79 @@
...
+        while (matcher.find()) {
+            result.add(matcher.group());
+        }
+        return result;
+    }
+
+    private void checkSlackNamesPresence() {
+        if (userCount == 0) {
+            logger.warn("The text: [{}] doesn't contain slack name.");
+            throw new WrongCommandFormatException(String.format("The text '%s' doesn't contains slackName", text));
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: contain without 's'

File: src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java [hunk 2]
```diff
@@ -1,10 +1,103 @@
...
+    @Test
+    public void getAllUsers() {
+        //given
+        String text = "text @slack2 text@slack1 text @slack3";
+        SlackParsedCommand slackParsedCommand = new SlackParsedCommand("@from", text, users);
+        //when
+        List<UserDTO> result = slackParsedCommand.receiveAllUsers();
+        //then
+        assertEquals("[UserDTO(uuid=uuid2, slack=@slack2), UserDTO(uuid=uuid1, slack=@slack1)]",
+                result.toString());
```
Comment (NikolayNN) [inline src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java]: The text "text @slack2 text@slack1 text @slack3" has three slackNames, but result has only two Users

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 3]
```diff
@@ -1,19 +1,79 @@
...
- * @author
+ * @author Konstantin Sergey
  */
 public class SlackParsedCommand {
-   private String from;
-   private String text;
-   private Map<String, UserDTO> users;
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    private final String SLACK_NAME_PATTERN = "@([a-zA-z0-9\\.\\_\\-]){1,21}";
+    private String from;
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: rename the field to fromSlackName

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 4]
```diff
@@ -1,19 +1,79 @@
...
+        if (!from.startsWith("@")) {
+            logger.debug("add '@' to slack name [{}]", from);
+            from = "@" + from;
+        }
         this.from = from;
         this.text = text;
         this.users = users;
-        //todo task #11
+        slackNames = receiveSlackNames();
+        userCount = slackNames.size();
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: add logger.debug here with text "SlackParsedCommand created. fromSlackName: {} text: {} slackNames in the text {slackNames.toString} users: {users.toString}"
or you can do better

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 5]
```diff
@@ -1,19 +1,79 @@
...
+    private Map<String, UserDTO> users;
+    private List<String> slackNames;
+    private int userCount;
 
     public SlackParsedCommand(String from, String text, Map<String, UserDTO> users) {
+        if (!from.startsWith("@")) {
+            logger.debug("add '@' to slack name [{}]", from);
+            from = "@" + from;
+        }
         this.from = from;
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: rename to fromSlackName

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 6]
```diff
@@ -1,19 +1,79 @@
...
+        userCount = slackNames.size();
+    }
+
+    public List<UserDTO> receiveAllUsers() {
+        checkSlackNamesPresence();
+        Map<String, UserDTO> usersCopy = new HashMap<>();
+        usersCopy.putAll(users);
+        usersCopy.remove(from);
+        logger.debug("Found {} users in text: [{}]", userCount, text);
+        return new LinkedList(usersCopy.values());
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: We have recently decided to output more information in a debug log. And I think that we should output users from the LinkedList to debug log instead userCount.

**Review by NikolayNN — CHANGES_REQUESTED** [review]

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 7]
```diff
@@ -1,19 +1,79 @@
...
 public class SlackParsedCommand {
-   private String from;
-   private String text;
-   private Map<String, UserDTO> users;
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    private final String SLACK_NAME_PATTERN = "@([a-zA-z0-9\\.\\_\\-]){1,21}";
+    private String from;
+    private String text;
+    private Map<String, UserDTO> users;
+    private List<String> slackNames;
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: I think better name to the field slackNamesInText

**Review by NikolayNN — CHANGES_REQUESTED** [review]

File: src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java (hunk 2 above)
Comment (ksergey12) [reply src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java]: Fixed.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 3 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Done.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 5 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Done.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 4 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Done.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 6 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Done.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 1 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Fixed.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 7 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Fixed.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 8]
```diff
@@ -1,19 +1,88 @@
...
         this.users = users;
-        //todo task #11
+        slackNamesInText = receiveSlackNames();
+        userCount = slackNamesInText.size();
+        logger.debug("SlackParsedCommand created with parameters: " +
+                "fromSlackName: {} text: {} slackNamesInText in the text {} users: {}",
+                fromSlackName, text, slackNamesInText.toString(), users.toString());
+    }
+
+    public List<UserDTO> receiveAllUsers() {
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: rename to getAllUsers

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 9]
```diff
@@ -1,19 +1,88 @@
...
+        checkSlackNamesPresence();
+        Map<String, UserDTO> usersCopy = new HashMap<>();
+        usersCopy.putAll(users);
+        usersCopy.remove(fromSlackName);
+        logger.debug("Found users {} in text: [{}]", users.toString(), text);
+        return new LinkedList(usersCopy.values());
+    }
+
+    public UserDTO receiveFirstUser() {
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: rename to getFirstUser

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 10]
```diff
@@ -1,19 +1,88 @@
...
+        return firstUser;
+    }
+
+    public String getTextWithoutSlackNames(){
+        String result = text.replaceAll(SLACK_NAME_PATTERN, "");
+        result = result.replaceAll("\\s+"," ").trim();
+        return result;
+    }
+
+    public UserDTO receiveFromUser() {
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: rename to getFromUser

**Review by NikolayNN — CHANGES_REQUESTED** [review]

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 8 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Done.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 9 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Done.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 10 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Done.

**Review by NikolayNN — APPROVED** [review]

**Review by OlgaKulikova — APPROVED** [review]

**Pull request merged**